### PR TITLE
lambdify loggamma now works for mpmath

### LINKED ISSUE #17411: lambdify loggamma doesn't work for mpmath
Lambdifying an expression with loggamma doesn't work for mpmath
```julia
In [6]: lambdify(x, loggamma(x), 'mpmath')                                                                                        
---------------------------------------------------------------------------
ImportError                               Traceback (most recent call last)
<ipython-input-6-02428e0a17ae> in <module>
----> 1 lambdify(x, loggamma(x), 'mpmath')

~/current/sympy/sympy/sympy/utilities/lambdify.py in lambdify(args, expr, modules, printer, use_imps, dummify)
    766                 imp_mod_lines.append("from %s import %s" % (mod, k))
    767     for ln in imp_mod_lines:
--> 768         exec_(ln, {}, namespace)
    769 
    770     # Provide lambda expression with builtins, and compatible implementation of range

<string> in <module>

ImportError: cannot import name 'lgamma'
```
I think this is because it's called loggamma rather than lgamma:
```julia
In [7]: import mpmath                                                                                                             

In [8]: mpmath.loggamma(1)                                                                                                        
Out[8]: mpf('0.0')
```

I originally got this error from
```julia
In [11]: nsolve(im(loggamma(I*x)), x, 3.439*I)                                                                                    
---------------------------------------------------------------------------
ImportError
...
```

### PR DESCRIPTION
**Pull request by abhaydhiman**
#### References to other Issues or PRs

Fixed #17411

#### Brief description of what is fixed or changed
Lambdifying an expression with loggamma doesn't work for mpmath previously but I made some fixes so that it works. File changed are pycode and test_pycode.

#### Other comments
Nan

#### Release Notes

* printing
    * The mpmath code printer now correctly prints the `loggamma` function.
* utilities
    * Lambdifying an expression with `loggamma` using mpmath no longer raises `ImportError`.

### CODE REVIEW

File: sympy/utilities/tests/test_lambdify.py
Comment (eric-wieser) [inline sympy/utilities/tests/test_lambdify.py:188]: Why is this needed?
```
Python 3.8.0 (tags/v3.8.0:fa919fd, Oct 14 2019, 19:37:50) [MSC v.1916 64 bit (AMD64)]
Type 'copyright', 'credits' or 'license' for more information
IPython 8.0.0.dev -- An enhanced Interactive Python. Type '?' for help.

In [1]: import sympy

In [2]: 'loggamma' in sympy.__dict__
Out[2]: True
```

**Comment by abhaydhiman** [comment]
As `sympy.__dict__` doesn't contains `lgamma`. Its used to pass the test case for `lgamma` as main function for it is defined in `mpmath.__dict__`.

File: sympy/utilities/lambdify.py
Comment (eric-wieser) [inline sympy/utilities/lambdify.py:79]: Are you sure you don't have this backwards?

```suggestion
    "loggamma": "lgamma",
```

Comment (abhaydhiman) [reply sympy/utilities/lambdify.py:79]: But then `assert mat in mpmath.__dict__` this test case would fail as `mpmath` also not contains `lgamma`. Now another change would like be `"loggamma": "loggamma"` but this doesn't make any sense according to me. The main point over here is that for `lgamma` the function of `loggamma` should be called that is available in `mpmath.__dict__` and thats exactly what I have done.

Comment (eric-wieser) [reply sympy/utilities/lambdify.py:79]: What is `lgamma`? If it's not a sympy function, why is it on this dict?

Comment (abhaydhiman) [reply sympy/utilities/lambdify.py:79]: Because in the related issue it is showing the import error for `lgamma` and It should import `loggamma` when it is showing that error message thats the only reason for adding `lgamma` in the dict. Further computation can be done by `loggamma` itself.

Comment (abhaydhiman) [reply sympy/utilities/lambdify.py:79]: Any suggested change?

Comment (eric-wieser) [reply sympy/utilities/lambdify.py:79]:
```suggestion
    "loggamma":"loggamma",
```
Assuming that also works. It results in a simpler test, and is what a maintainer suggested in the original issue.

Comment (abhaydhiman) [reply sympy/utilities/lambdify.py:79]: Ok, I am on it.

Comment (abhaydhiman) [reply sympy/utilities/lambdify.py:79]: If I try 
```
from sympy import *
import mpmath

x = symbols('x')
print(lambdify(x, loggamma(x), 'mpmath'))
```
with `"loggamma": "loggamma"` I still got this error:



```
Traceback (most recent call last):
  File "E:\sympy\jupyter_files\sympy\sympy\utilities\lambdify.py", line 835, in lambdify
    exec_(ln, {}, namespace)
  File "<string>", line 1, in <module>
ImportError: cannot import name 'lgamma' from 'mpmath' (E:\sympy\jupyter_files\test\lib\site-packages\mpmath\__init__.py)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "E:\sympy\jupyter_files\sympy\sympy\utilities\lambdify.py", line 841, in lambdify
    exec_(ln, {}, namespace)
  File "<string>", line 1, in <module>
NameError: name 'mpmath' is not defined
```
But when `"lgamma": "loggamma"` I got the generated function like this:-
```
<function _lambdifygenerated at 0x00000184C93077B8>
```

Comment (abhaydhiman) [reply sympy/utilities/lambdify.py:79]: In `mpmath __init__.py ` if I add `lgamma = mp.loggamma` then also it works and showing the correct output.

Comment (eric-wieser) [reply sympy/utilities/lambdify.py:79]: Thanks for trying that out. Something weird is going on here.

Comment (abhaydhiman) [reply sympy/utilities/lambdify.py:79]: > Something weird is going on here.

Absolutely I am also wondering that why its tries to import `lgamma`.

Comment (eric-wieser) [reply sympy/utilities/lambdify.py:79]: My guess is that somewhere a translation from `sympy.loggamma` to  `math.lgamma` happens, and then your patch reverses this mistake.

Comment (eric-wieser) [reply sympy/utilities/lambdify.py:79]: Ah, found it: these lines are incorrect: https://github.com/sympy/sympy/blob/70f83bd004e35fc3b96a1e07699a3db33a84002c/sympy/printing/pycode.py#L499

Comment (abhaydhiman) [reply sympy/utilities/lambdify.py:79]: > My guess is that somewhere a translation from `sympy.loggamma` to `math.lgamma` happens, and then your patch reverses this mistake.

Yup

**Comment by eric-wieser** [comment]
You need to change the mpmath specific lines I linked to above.

**Comment by abhaydhiman** [comment]
> You need to change the mpmath specific lines I linked to above.

What exactly do I have to change in that line of code.

**Comment by eric-wieser** [comment]
Updating the dictionary to correct the lgamma key on the next line would probably do the trick.

**Comment by abhaydhiman** [comment]
well, I actually did that.

File: sympy/printing/pycode.py
Comment (eric-wieser) [inline sympy/printing/pycode.py:43]: This is used for the math module, so should not be changed here.

Comment (abhaydhiman) [reply sympy/printing/pycode.py:43]: ok ok

**Comment by eric-wieser** [comment]
You should add to `_known_functions_mpmath`

**Comment by abhaydhiman** [comment]
`'lgamma': 'loggamma'` if I add this in `_known_functions_mpmath` that again gives the same error.

**Comment by eric-wieser** [comment]
> Or I have to add `"loggamma": "loggamma"` without deleting the previous `"loggamma": "lgamma"`, I tried this and this works perfectly fine.

This sounds ideal to me - this effectively _does_ delete the previous entry.

**Comment by abhaydhiman** [comment]
Is it good to go now?

**Comment by eric-wieser** [comment]
No. You need to add `"loggamma": "loggamma"` to `_known_functions_mpmath`

**Comment by abhaydhiman** [comment]
oh! Sorry, I am on it.

**Comment by eric-wieser** [comment]
This fix now looks correct - can you add a test?

**Comment by abhaydhiman** [comment]
Testcase for this ` nsolve(im(loggamma(I*x)), x, 3.439*I)` equation?

**Comment by abhaydhiman** [comment]
Also, do I have to add a test in `test_pycode.py`?

**Comment by eric-wieser** [comment]
Yes, please add a test to that file. Simply checking the result of something like `pycode(loggamma(x))` should be enough, no need to complicate the test

**Comment by abhaydhiman** [comment]
Ok Thanks, can I make a separate function for that?

**Comment by abhaydhiman** [comment]
Or I have to add that in `test_PythonCodePrinter`

**Comment by eric-wieser** [comment]
`test_MpmathPrinter` looks like a good place to add it, since this is about mpmath.

**Review by eric-wieser — APPROVED** [review]
Release notes need updating to describe what was fixed (and not how), but the code is good.

**Comment by abhaydhiman** [comment]
But it got AssertionError. I don't know why that's happen

File: sympy/printing/tests/test_pycode.py
Comment (eric-wieser) [inline sympy/printing/tests/test_pycode.py:79]:
```suggestion
    assert p.doprint(loggamma(x)) == 'mpmath.loggamma(x)'
```

**Comment by abhaydhiman** [comment]
I have updated the release notes please review it. Thanks!!

**Comment by eric-wieser** [comment]
Still not quite right - will defer to @oscarbenjamin, since they reported the original bug.

**Comment by oscarbenjamin** [comment]
It looks like this fixes the original issue. What is not quite right?

Is it the release note?

Maybe it should say:
```
* printing
    * The mpmath code printer now correctly prints the loggamma function.
* utilities
    * Lambdifying an expression with loggamma using mpmath no longer raises NameError.
```

**Comment by eric-wieser** [comment]
Yeah, just the release note which decribed the diff not the result. That looks like a good note to me.

**Comment by eric-wieser** [comment]
Although I think it was ImportError

**Comment by oscarbenjamin** [comment]
Maybe there could also be a test for lambdify

**Comment by abhaydhiman** [comment]
So do I have to include a test for lambdify also?

**Comment by abhaydhiman** [comment]
@oscarbenjamin or @eric-wieser Do I have to add Test in `test_lambdify.py` file specifically in `test_special_printers` function if not please guide me where I have to add the Testcase for lambdify?

**Comment by abhaydhiman** [comment]
Is this `lambdify(x, loggamma(x), 'mpmath')(5)` test is valid?

**Comment by oscarbenjamin** [comment]
Yes, that seems good. You could test that it gives a couple of expected values for the loggamma function.

**Comment by abhaydhiman** [comment]
Yes, tried different values for x and it gives the expected output. So Is one test case is sufficient or I have to add more than one test for lambdify.

**Comment by abhaydhiman** [comment]
@oscarbenjamin I have added the test case please review it.

**Comment by oscarbenjamin** [comment]
Looks good

**Comment by abhaydhiman** [comment]
Thanks!

**Comment by abhaydhiman** [comment]
@oscarbenjamin any other change, you wanted in this PR.

**Review by Ayush-Malik — APPROVED** [review]
Looks good to me , @oscarbenjamin  please review this PR

**Comment by czgdp1807** [comment]
@oscarbenjamin Is this good to go?

**Comment by oscarbenjamin** [comment]
Looks good. Thanks!

**Pull request merged**
